packages/query-core/src/tests/infiniteQueryBehavior.test.tsx

```diff
@@ -121,4 +121,138 @@ describe('InfiniteQueryBehavior', () => {
 
     unsubscribe()
   })
+
+  test('InfiniteQueryBehavior should apply the maxPages option to limit the number of pages', async () => {
+    const key = queryKey()
+    let abortSignal: AbortSignal | null = null
+
+    const queryFnSpy = jest
+      .fn()
+      .mockImplementation(({ pageParam = 1, signal }) => {
+        abortSignal = signal
+        return pageParam
+      })
+
+    const observer = new InfiniteQueryObserver<number>(queryClient, {
+      queryKey: key,
+      queryFn: queryFnSpy,
+      getNextPageParam: (lastPage) => lastPage + 1,
+      getPreviousPageParam: (firstPage) => firstPage - 1,
+      maxPages: 2,
+    })
+
+    let observerResult:
+      | InfiniteQueryObserverResult<unknown, unknown>
+      | undefined
+
+    const unsubscribe = observer.subscribe((result) => {
+      observerResult = result
+    })
+
+    // Wait for the first page to be fetched
+    await waitFor(() =>
+      expect(observerResult).toMatchObject({
+        isFetching: false,
+        data: { pages: [1] },
+      }),
+    )
+
+    expect(queryFnSpy).toHaveBeenNthCalledWith(1, {
+      queryKey: key,
+      pageParam: undefined,
+      meta: undefined,
+      signal: abortSignal,
+    })
+
+    queryFnSpy.mockClear()
+
+    // Fetch the second page
+    await observer.fetchNextPage()
+
+    expect(queryFnSpy).toHaveBeenNthCalledWith(1, {
+      queryKey: key,
+      pageParam: 2,
+      meta: undefined,
+      signal: abortSignal,
+    })
+
+    expect(observerResult).toMatchObject({
+      isFetching: false,
+      data: { pages: [1, 2] },
+    })
+
+    queryFnSpy.mockClear()
+
+    // Fetch the page before the first page
+    await observer.fetchPreviousPage()
+
+    expect(queryFnSpy).toHaveBeenNthCalledWith(1, {
+      queryKey: key,
+      pageParam: 0,
+      meta: undefined,
+      signal: abortSignal,
+    })
+
+    // Only first two pages should be in the data
+    expect(observerResult).toMatchObject({
+      isFetching: false,
+      data: { pages: [0, 1] },
+    })
+
+    queryFnSpy.mockClear()
+
+    // Fetch the page before
+    await observer.fetchPreviousPage()
+
+    expect(queryFnSpy).toHaveBeenNthCalledWith(1, {
+      queryKey: key,
+      pageParam: -1,
+      meta: undefined,
+      signal: abortSignal,
+    })
+
+    expect(observerResult).toMatchObject({
+      isFetching: false,
+      data: { pages: [-1, 0] },
+    })
+
+    queryFnSpy.mockClear()
+
+    // Fetch the page after
+    await observer.fetchNextPage()
+
+    expect(queryFnSpy).toHaveBeenNthCalledWith(1, {
+      queryKey: key,
+      pageParam: 1,
+      meta: undefined,
+      signal: abortSignal,
+    })
+
+    expect(observerResult).toMatchObject({
+      isFetching: false,
+      data: { pages: [0, 1] },
+    })
+
+    queryFnSpy.mockClear()
+
+    // Refetch the infinite query
+    await observer.refetch()
+
+    // Only 2 pages should be refetched
+    expect(queryFnSpy).toHaveBeenNthCalledWith(1, {
```

Comment:
does this really check if we have called it only 2 times? If I understand `toHaveBeenNthCalledWith`, it only checks if the 1st call has been called with the certain params and the 2nd call as well. But there could be a 3d call as well and this test would pass.

I think we should add:

`expect(queryFnSpy).toHaveBeenCalledTimes(2)`

Reply:
Good catch! I'll do.

Reply:
Done in 156751db8643d3cb477e9050df48b3d664f44012